src/Operators/spectralelement.jl

```diff
@@ -631,6 +631,156 @@ Base.@propagate_inbounds function apply_operator(
+"""
+struct SplitDivergence{I} <: SpectralElementOperator{I} end
+SplitDivergence() = SplitDivergence{()}()
+SplitDivergence{()}(space) = SplitDivergence{operator_axes(space)}()
+
+operator_return_eltype(::SplitDivergence{I}, ::Type{S1}, ::Type{S2}) where {I, S1, S2} =
+    RecursiveApply.rmaptype(Geometry.divergence_result_type, S1)
+
+function apply_operator(op::SplitDivergence{(1,)}, space, slabidx, arg1, arg2)
+    FT = Spaces.undertype(space)
+    QS = Spaces.quadrature_style(space)
+    Nq = Quadratures.degrees_of_freedom(QS)
+    D = Quadratures.differentiation_matrix(FT, QS)
+    RT = operator_return_eltype(op, eltype(arg1), eltype(arg2))
+    out = IF{RT, Nq}(MArray, FT)
+    fill!(parent(out), zero(FT))
+
+    Ju1_slab = MArray{Tuple{Nq}, FT}(undef)
+    psi_slab = MArray{Tuple{Nq}, FT}(undef)
+
+    @inbounds for i in 1:Nq
+        ij = CartesianIndex((i,))
+        local_geometry = get_local_geometry(space, ij, slabidx)
+        u = get_node(space, arg1, ij, slabidx)
+        psi = get_node(space, arg2, ij, slabidx)
+        Ju1 = local_geometry.J * Geometry.contravariant1(u, local_geometry)
+        Ju1_slab[i] = Ju1
+        psi_slab[i] = psi
+    end
+
+    @inbounds for i in 1:Nq
+        val = zero(FT)
+        for j in 1:Nq
+            flux_ij = 0.5 * (Ju1_slab[i] + Ju1_slab[j]) * (psi_slab[i] + psi_slab[j])
+            val += D[i, j] * flux_ij
+        end
+        out[slab_index(i)] = val
+    end
+
+    @inbounds for i in 1:Nq
+        ij = CartesianIndex((i,))
+        local_geometry = get_local_geometry(space, ij, slabidx)
+        out[slab_index(i)] = out[slab_index(i)] * local_geometry.invJ
+    end
+
+    return Field(SArray(out), space)
+end
+
+function apply_operator(op::SplitDivergence{(1, 2)}, space, slabidx, arg1, arg2)
+    FT = Spaces.undertype(space)
+    QS = Spaces.quadrature_style(space)
+    Nq = Quadratures.degrees_of_freedom(QS)
+    D = Quadratures.differentiation_matrix(FT, QS)
+    RT = operator_return_eltype(op, eltype(arg1), eltype(arg2))
+    out = DataLayouts.IJF{RT, Nq}(MArray, FT)
+    fill!(parent(out), zero(FT))
+
+    Ju1_slab = MArray{Tuple{Nq, Nq}, FT}(undef)
+    Ju2_slab = MArray{Tuple{Nq, Nq}, FT}(undef)
+    psi_slab = MArray{Tuple{Nq, Nq}, FT}(undef)
+
+    @inbounds for j in 1:Nq, i in 1:Nq
+        ij = CartesianIndex((i, j))
+        local_geometry = get_local_geometry(space, ij, slabidx)
+        u = get_node(space, arg1, ij, slabidx)
+        psi = get_node(space, arg2, ij, slabidx)
+
+        Ju1_slab[i, j] = local_geometry.J * Geometry.contravariant1(u, local_geometry)
+        Ju2_slab[i, j] = local_geometry.J * Geometry.contravariant2(u, local_geometry)
+        psi_slab[i, j] = psi
+    end
+
+    @inbounds for j in 1:Nq
+        for i in 1:Nq
+            val = zero(FT)
+            for k in 1:Nq
+                flux_ik =
+                    0.5 * (Ju1_slab[i, j] + Ju1_slab[k, j]) *
+                    (psi_slab[i, j] + psi_slab[k, j])
```

Comment:
This line and the corresponding one for `flux_jk` have the same `Float32`-`Float64` type instability, and they should also probably use values of `J` from different nodes.

Reply:
Fixed type instability